pytket_qir/gatesets/base.py

```diff
@@ -45,60 +48,17 @@ class OpSpec(Enum):
     CTLADJ = "ctladj"
 
 
-QirGate = NamedTuple(
-    "QirGate",
-    [
-        (
-            "opnat",
-            OpNat,
-        ),
-        (
-            "opname",
-            OpName,
-        ),
-        (
-            "opspec",
-            OpSpec,
-        ),
-    ],
-)
-
-
-CustomQirGate = NamedTuple(
-    "CustomQirGate",
-    [
-        (
-            "opnat",
-            OpNat,
-        ),
-        (
-            "opname",
-            OpName,
-        ),
-        (
-            "opspec",
-            OpSpec,
-        ),
-        (
-            "function_signature",
-            List[PyQirInputTypes],
-        ),
-        (
-            "return_type",
-            PyQirOutputTypes,
-        ),
-    ],
-)
+@dataclass(frozen=True)
+class QirGate:
+    opnat: OpNat
+    opname: Enum
```

Comment:
Yes, this is one bit that I'd better find another solution. In brief, initially, `opnat`, `opname` and `opspec` were enums hence immutable. But it turned out while working on including WASM boxes that I needed mutability to define a `QirGate` object for it at runtime and especially the function name. This is why I re-define an enum to hold that name [here](https://github.com/CQCL/pytket-qir/blob/079202fe029c989171a622483421b73cd1e9319b/pytket_qir/qir.py#L488). Therefore, in `QirGate`, `opname` is defined as a generic `Enum` type. I'd be happy to hear another solution as I am not convinced this is the best one.

Reply:
I don't quite understand the problem is here. What goes wrong if you make `opname` a string?

Reply:
Well, the issue is that WASM function calls are defined as external function calls in `pyqir-generator` (specifically [here](https://github.com/CQCL/pytket-qir/blob/9ed1c28417f44c8b741d55df23502fd1f8a5af60/pytket_qir/qir.py#L525)). To define them, I have to add ir to the `pyqir` module from an input gateset (teh machinery [here](https://github.com/CQCL/pytket-qir/blob/9ed1c28417f44c8b741d55df23502fd1f8a5af60/pytket_qir/qir.py#L376)). Previously, since the gateset was user defined and inputted alongside the circuit, it didn't need to be mutated at runtime. Now, when adding WASM calls, I need to update the gateset when processing the WASMOp, which means creating a new `QirGate` ([here](https://github.com/CQCL/pytket-qir/blob/9ed1c28417f44c8b741d55df23502fd1f8a5af60/pytket_qir/qir.py#L503)), updating the gateset ([here](https://github.com/CQCL/pytket-qir/blob/9ed1c28417f44c8b741d55df23502fd1f8a5af60/pytket_qir/qir.py#L509)) and finally updating the module ([here](https://github.com/CQCL/pytket-qir/blob/9ed1c28417f44c8b741d55df23502fd1f8a5af60/pytket_qir/qir.py#L518)). This was impossible with the previously defined datatypes (`OpName`, `OpType` etc) that were all enums. But in truth, the only one that needs to be updated is the `opname` field in `QirGate`. This is why I defined it as a generic `Enum` type here.

Reply:
Can we have a specific enum of op types, which includes the WASM op? It would be more satisfying if this were captured as a single enum rather than a generic enum type, which could be anything at all.

Reply:
I came up with a slightly better solution but I couldn't get away with creating a singleton enum on the fly for the purposes of holding the WASM Function name. This is because all the machinery for defining external function calls rely on enums syntax and it is really tedious to adapt that for a specialised case where the gate definition will be of another (more flexible?) type.

Reply:
Well, no more luck. To comply with mypy, I have to define a union of types with the generic `Enum`. I don't really see any way around this, sorry.